lib/ddtrace/contrib/rails/core_extensions.rb

```diff
@@ -1,50 +1,124 @@
-            render_template_without_datadog(*args, &block)
+    def patch_template_renderer(klass)
+      klass.class_eval do
+        def render_with_datadog(*args, &block)
+          # create a tracing context and start the rendering span
+          # NOTE: Rails < 3.1 compatibility: preserve the tracing
+          # context when a partial is rendered
+          @tracing_context ||= {}
+          if @tracing_context.empty?
+            ::ActiveSupport::Notifications.instrument(
+              '!datadog.start_render_template.action_view',
+              tracing_context: @tracing_context
+            )
           end
+          render_without_datadog(*args)
+        rescue Exception => e
+          # attach the exception to the tracing context if any
+          @tracing_context[:exception] = e
+          raise e
+        ensure
+          # ensure that the template `Span` is finished even during exceptions
+          ::ActiveSupport::Notifications.instrument(
+            '!datadog.finish_render_template.action_view',
+            tracing_context: @tracing_context
+          )
         end
-      else # Rails < 3.1
-        ::ActionView::Template.class_eval do
-          alias_method :render_template_without_datadog, :render
-          def render(*args, &block)
-            ActiveSupport::Notifications.instrument('start_render_template.action_view')
-            render_template_without_datadog(*args, &block)
-          end
+
+        def render_template_with_datadog(*args)
+          # arguments based on render_template signature (stable since Rails 3.2)
+          template = args[0]
+          layout_name = args[1]
+
+          # update the tracing context with computed values before the rendering
+          template_name = template.try('identifier')
+          template_name = Datadog::Contrib::Rails::Utils.normalize_template_name(template_name)
+          layout = if layout_name.is_a?(String)
+                     # NOTE: Rails < 3.1 compatibility: the second argument is the layout name
+                     layout_name
+                   else
+                     layout_name.try(:[], 'virtual_path')
+                   end
+          @tracing_context[:template_name] = template_name
+          @tracing_context[:layout] = layout
+        rescue StandardError => e
+          Datadog::Tracer.log.error(e.message)
+        ensure
+          render_template_without_datadog(*args)
+        end
+
+        # method aliasing to patch the class
+        alias_method :render_without_datadog, :render
+        alias_method :render, :render_with_datadog
+
+        if klass.private_method_defined?(:render_template) || klass.method_defined?(:render_template)
+          alias_method :render_template_without_datadog, :render_template
+          alias_method :render_template, :render_template_with_datadog
+        else
+          # NOTE: Rails < 3.1 compatibility: the method name is different
```

Comment:
Nitpicking, I don't understand your comment with GH then -> if `Rails >= 4` ||  `Rails == 3.2` else `Rails < 3.2`, because this inline comment says `Rails < 3.1` (or should it be `Rails <= 3.1` ?)

Reply:
oh yes sorry, basically was to say:
* `klass.private_method_defined?(:render_template)` is the case of Rails >=4
* `klass.method_defined?(:render_template)` is the case of Rails == 3.2
* and the else is Rails < 3.2 (not 3.1, I may copied that comment from somewhere else)